app/utils/ServiceWatcher.js

```diff
@@ -0,0 +1,104 @@
+const {
+  SUBSTRATE_STATUS_POLL_PERIOD_MS,
+  SUBSTRATE_STATUS_TIMEOUT_MS
+} = require('../env')
+
+class ServiceWatcher {
+  // taking helper functions can be improved by taking an object
+  // { name: <srv_name>, method: createNodeApi } and use in init()
+  constructor(createNodeApi) {
+    this.report = {}
+    this.pollPeriod = SUBSTRATE_STATUS_POLL_PERIOD_MS
+    this.timeout = SUBSTRATE_STATUS_TIMEOUT_MS
+    this.createNodeApi = createNodeApi
+  }
+
+  // substrate polling function, each service should have their own
+  async #substratePoll({ createNodeApi, name = 'substrate' }) {
+    try {
+      const api = (await createNodeApi())._api
+      if (!await api.isReady) throw new Error('service is not ready')
+      const [chain, runtime] = await Promise.all([api.runtimeChain, api.runtimeVersion])
+    
+      return {
+        name,
+        status: 'up', // TODO repl with constant/symbol 
+        details: {
+          chain,
+          runtime: {
+            name: runtime.specName,
+            versions: {
+              spec: runtime.specVersion, //.toNumber(), // tmp commenting out for stubbing
+              impl: runtime.implVersion, //.toNumber(),
+              authoring: runtime.authoringVersion, //.toNumber(),
+              transaction: runtime.transactionVersion //.toNumber(),
+            },
+          },
+        }
+      } 
+    } catch(error) {
+      // TODO logging
+      return { name, status: 'error', error }
+    }
+  }
```

Comment:
I don't like how this conflates the management of the polling cycle with the specifics of this check. Can this be refactored to inject the dependency or perhaps use inheritance to implement multiple checks?

Reply:
gotcha

Reply:
done